### ISSUE TITLE
[Potential contribution] Minibatch trimming (curriculum learning method)

### ISSUE THREAD

**Issue by hfassold** [microsoft/DeepSpeed#1539]
I would like to contribute an easy implementable curriculum learning method ("minibatch trimming", see https://github.com/hfassold/mbtrim ) to the framework.
Is that possible ? What are the steps I have to do in order to do that ?

**Comment by conglongli** [comment]
Hi @hfassold thank you for your interest in contributing to DeepSpeed. Because we didn't receive this kind of contribution before (a brand new technique that could affect training convergence), we are having some internal discussions about the general contribution guideline and haven't made a conclusion yet. We will let you know when we have a guideline ready. Thank you for your patience.

**Comment by hfassold** [comment]
Alright, no problem, I have time. Just inform me when you have setup the guidelines.

**Comment by conglongli** [comment]
Hi @hfassold we have just released the new feature contribution guideline: https://github.com/microsoft/DeepSpeed/blob/master/CONTRIBUTING.md#new-feature-contribution-guidelines

For your case, our major concern is that there is no evaluation on any transformer model and on any NLP task, which are important for many DeepSpeed users. If you are able to add these necessary evaluations, we are happy to take another look. Thank you!

**Comment by hfassold** [comment]
Sounds good, will do. Will  come back to you then (in january I suppose)

**Comment by hfassold** [comment]
I made now a test of minibatch-trimming for fine-tuning a transformer model for an NLP task.
I adapted the code from:
https://github.com/hfwittmann/transformer_finetuning_lightning
I took this code because it is uses Pytorch Lightning (which I prefer to 'pure' Pytorch).
Furthermore, it is quite compact and good to understand (not that I am more on the vision side, I just started with NLP models/frameworks)
It does fine-tuning of an NLP model (I use a pretrained 'DistilliBert'  network) on the task of text classification.
Specifically, the task is to predict the rating (1 - 5 stars) given by a certain user for a Android App, based on his review text.

I do fine-tuning for 10 epochs, with 'AdamW' algorithm.
As a performance measure, I take the best validation accuracy during the last 5 epochs of training.
I do 10 runs, and calculate the average accuracy from these 10 runs.
For mini-batch trimming I use the default parameters (with _p_ decreasing linearly from 1.0 to 0.2) which I use also for vision tasks.

The results are now as follows:
Average val. accuracy with mini-batch trimming _disabled_: **79,94 %**
Average val. accuracy with mini-batch trimming _enabled_: **79,84 %**
One can interpret this result either as 'on par' or as a slight regression, I will leave that up to you.
From my side, I would say this means that mini-batch trimming is not that effective for fine-tuning a NLP model than for training from scratch a vision model (where it definitly brings a benefit as can be seen in my ASPAI paper, at https://arxiv.org/abs/2110.13058).

It's now up to you to decide whether it's worth to integrate into DeepSpeed, or not.
From my side, I would be happy to integrate it.
But currently I am quite busy, so it would be (from my side) something for Q2 2022.

**Comment by conglongli** [comment]
Hi @hfassold thank you for sharing your investigation, which is very helpful. On our side, we are actually also investigating some ways to manipulate the losses in order to improve the training convergence since last summer. We are studying some different strategies that we feel would be more fit for NLP area, but we haven't made any conclusion yet.

Given that currently you won't have time to implement your solution, my proposal is that we can revisit this later. If later we got some positive results in NLP area, I feel that it might make sense on our side to implement both yours and ours solutions together (will definitely give credit to your work) as a single whole feature. Just wonder what do you think.

Regarding the optimizer direction, we do have similar interests. It'd be helpful if you could open a separate issue about it when your paper is ready. Thank you.

**Comment by hfassold** [comment]
Sounds like a good plan. Curriculum learning seems definitly be beneficial, especially for (pre)training 'from scratch'. For finetuning, I am not so sure. OK, I will wait until you have some conclusions regarding the best strategy. Please inform me then (either her on this thread, or via PM to hannes.fassold@joanneum.at). And yes, it makes definitly sense to implement this together as a single whole feature.

Regarding the 'AdaFamily' optimizer, yes I will open a separate issue then. As mentioned, the results look good.

**Comment by conglongli** [comment]
Sounds good, yes we will keep you posted.

**Comment by hfassold** [comment]
Regarding 'AdaFamily' - will put the preprint onto arxiv, and then I will open a separate '[Potential contribution]' issue begin of next week

**Comment by hfassold** [comment]
I found now the reason why I do NOT observed a speedup during training, with mini-batch trimming enabled.
Seems to be that pytorch is NOT optimized for the case when forward and backward pass have _different_ batch sizes.
See the pytorch issue I opened at https://github.com/pytorch/pytorch/issues/74604

**Comment by conglongli** [comment]
Closing this as we introduced curriculum learning/data efficiency features that provide decent improvements and verified on compute-heavy NLP/CV tasks https://www.deepspeed.ai/2022/12/11/data-efficiency.html. But users can still apply similar kind of weighted loss on user-side code when using DeepSpeed.
